cpp/benchmarks/fixture/nvbench_fixture.hpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2021-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/mr/device/arena_memory_resource.hpp>
+#include <rmm/mr/device/cuda_async_memory_resource.hpp>
+#include <rmm/mr/device/cuda_memory_resource.hpp>
+#include <rmm/mr/device/managed_memory_resource.hpp>
+#include <rmm/mr/device/owning_wrapper.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+#include <string>
+
+namespace cudf {
+namespace detail {
+static std::string rmm_mode_param{"--rmm_mode"};  ///< RMM mode command-line parameter name
+}  // namespace detail
+
+/**
+ * Base fixture for cudf benchmarks using nvbench.
+ *
+ * Initializes the default memory resource to use the RMM pool device resource.
+ */
+struct nvbench_base_fixture {
+  inline auto make_cuda() { return std::make_shared<rmm::mr::cuda_memory_resource>(); }
+
+  inline auto make_pool()
+  {
+    return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(make_cuda());
+  }
+
+  inline auto make_async() { return std::make_shared<rmm::mr::cuda_async_memory_resource>(); }
+
+  inline auto make_managed() { return std::make_shared<rmm::mr::managed_memory_resource>(); }
+
+  inline auto make_arena()
+  {
+    return rmm::mr::make_owning_wrapper<rmm::mr::arena_memory_resource>(make_cuda());
+  }
+
+  inline std::shared_ptr<rmm::mr::device_memory_resource> create_memory_resource(
+    std::string const& mode)
+  {
+    if (mode == "cuda") return make_cuda();
+    if (mode == "pool") return make_pool();
+    if (mode == "async") return make_async();
+    if (mode == "arena") return make_arena();
+    if (mode == "managed") return make_managed();
+    CUDF_FAIL("unknown rmm_mode parameter: " + mode);
+  }
+
+  nvbench_base_fixture(int argc, char const* const* argv)
+  {
+    for (int i = 1; i < argc - 1; ++i) {
+      std::string arg = argv[i];
+      if (arg == detail::rmm_mode_param) {
+        i++;
+        rmm_mode = argv[i];
+      }
```

Comment:
There is no checking here for things like `bench --rmm_mode cuda --rmm_mode pool`. I don't know if this is necessary to check though.

Reply:
I don't think this is worth checking for.

Reply:
The current behavior would select the last value provided when multiple `--rmm_mode` values exist. That matches the behavior of linux command line programs ( e.g. gcc, ls, ... )

Now what we are missing ( if we care ) is error handling around usage of `--rmm_mode` with no following value.

Reply:
It defaults to 'pool' just fine in this case. That was my intention in this situation.

Reply:
Ahh, you are correct. I missed that you are iterating to `argc-1` and therefore ignore a trailing `--rmm_mode` without a value. :+1: